fs/nova/balloc.c

```diff
@@ -574,6 +574,114 @@ struct nova_range_node *nova_alloc_blocknode_atomic(struct super_block *sb)
 	return nova_alloc_range_node_atomic(sb);
 }
 
+#define PAGES_PER_2MB 512
+#define PAGES_PER_2MB_MASK (512 - 1)
+#define IS_DATABLOCKS_2MB_ALIGNED(numblocks, atype) \
+		(!(num_blocks & PAGES_PER_2MB_MASK) && (atype == DATA))
+
+bool nova_alloc_superpage(struct super_block *sb,
+	struct free_list *free_list, unsigned long num_blocks,
+	unsigned long *new_blocknr, enum nova_alloc_direction from_tail)
+{
+	struct rb_root *tree;
+	struct rb_node *temp;
+	struct nova_range_node *curr;
+	unsigned long curr_blocks;
+	bool found = 0;
+	unsigned long step = 0;
+
+	unsigned int left_margin;
+	unsigned int right_margin;
+
+	tree = &(free_list->block_free_tree);
+	if (from_tail == ALLOC_FROM_HEAD)
+		temp = &(free_list->first_node->node);
+	else
+		temp = &(free_list->last_node->node);
+
+	while (temp) {
+		step++;
+		curr = container_of(temp, struct nova_range_node, node);
+
+		if (!nova_range_node_checksum_ok(curr)) {
+			nova_err(sb, "%s curr failed\n", __func__);
+			goto next;
+		}
+
+		curr_blocks = curr->range_high - curr->range_low + 1;
+		left_margin = PAGES_PER_2MB - (curr->range_low & PAGES_PER_2MB_MASK);
+
+		/*
+		 * Guard against cases where:
+		 * a. Unaligned free blocks is smaller than #512
+		 *    left_margin could larger than curr_blocks.
+		 * b. After alignment, free blocks is smaller than
+		 *    requested blocks.
+		 * Otherwise, we are free to go.
+		 */
+		if ((curr_blocks > left_margin) && \
+			(num_blocks <= (curr_blocks - left_margin))) {
+			struct nova_range_node *node;
+
+			right_margin = curr_blocks - left_margin - num_blocks;
+			nova_dbgv("curr:%p: num_blocks:%lu curr->range_low:%lu high:%lu",
+						curr, num_blocks, curr->range_low, curr->range_high);
+
+			/*
+			 * Don't overlap with existing node,
+			 * Free curr before insert the left,right margin node.
+			 */
+			rb_erase(&curr->node, tree);
```

Comment:
Please reuse the curr node, just change its range_low/range_high and checksum so that we can remove one node free and one allocation.